// openapi-2-kong/dist/types/kong.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xKongOpenIDConnect = exports.xKongBasicAuth = exports.xKongPluginRequestTermination = exports.xKongPluginKeyAuth = exports.xKongPluginRequestValidator = exports.isParameterSchema = exports.isBodySchema = exports.xKongServiceDefaults = exports.xKongUpstreamDefaults = exports.xKongRouteDefaults = exports.xKongName = void 0;
exports.xKongName = 'x-kong-name';
exports.xKongRouteDefaults = 'x-kong-route-defaults';
exports.xKongUpstreamDefaults = 'x-kong-upstream-defaults';
exports.xKongServiceDefaults = 'x-kong-service-defaults';
const isBodySchema = (value = {}) => (Object.prototype.hasOwnProperty.call(value, 'body_schema') && value.body_schema !== undefined);
exports.isBodySchema = isBodySchema;
const isParameterSchema = (value = {}) => (Object.prototype.hasOwnProperty.call(value, 'parameter_schema') && value.parameter_schema !== undefined);
exports.isParameterSchema = isParameterSchema;
exports.xKongPluginRequestValidator = 'x-kong-plugin-request-validator';
exports.xKongPluginKeyAuth = 'x-kong-plugin-key-auth';
exports.xKongPluginRequestTermination = 'x-kong-plugin-request-termination';
exports.xKongBasicAuth = 'x-kong-plugin-basic-auth';
exports.xKongOpenIDConnect = 'x-kong-plugin-openid-connect';
