// openapi-2-kong/dist/declarative-config/jest/test-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSpec = exports.dummyPluginDoc = exports.dummyName = exports.pluginDummy = exports.tags = void 0;
exports.tags = ['Tag'];
exports.pluginDummy = {
    'x-kong-plugin-dummy': {
        name: 'dummy',
        config: {
            foo: 'bar',
        },
    },
};
const dummyName = (suffix) => `add-dummy-${suffix}`;
exports.dummyName = dummyName;
const dummyPluginDoc = (suffix) => ({
    apiVersion: 'configuration.konghq.com/v1',
    config: {
        foo: 'bar',
    },
    kind: 'KongPlugin',
    metadata: {
        name: exports.dummyName(suffix),
    },
    plugin: 'dummy',
});
exports.dummyPluginDoc = dummyPluginDoc;
/** This function is written in such a way as to allow mutations in tests but without affecting other tests. */
const getSpec = (overrides = {}) => JSON.parse(JSON.stringify({
    openapi: '3.0',
    info: {
        version: '1.0',
        title: 'My API',
    },
    servers: [
        {
            url: 'https://server1.com/path',
        },
    ],
    paths: {
        '/cats': {
            'x-kong-name': 'Cat stuff',
            summary: 'summary is ignored',
            post: {},
        },
        '/dogs': {
            summary: 'Dog stuff',
            get: {},
            post: {
                summary: 'Ignored summary',
            },
        },
        '/birds/{id}': {
            get: {},
        },
    },
    ...overrides,
}));
exports.getSpec = getSpec;
