// openapi-2-kong/dist/declarative-config/generate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDeclarativeConfigFromSpec = void 0;
const services_1 = require("./services");
const upstreams_1 = require("./upstreams");
function generateDeclarativeConfigFromSpec(api, tags) {
    try {
        const document = {
            _format_version: '1.1',
            services: services_1.generateServices(api, tags),
            upstreams: upstreams_1.generateUpstreams(api, tags),
        };
        const declarativeConfigResult = {
            type: 'kong-declarative-config',
            label: 'Kong Declarative Config',
            documents: [document],
            warnings: [],
        };
        // This removes any circular references or weirdness that might result from the JS objects used.
        // see: https://github.com/Kong/studio/issues/93
        const result = JSON.parse(JSON.stringify(declarativeConfigResult));
        return result;
    }
    catch (err) {
        throw new Error('Failed to generate spec: ' + err.message);
    }
}
exports.generateDeclarativeConfigFromSpec = generateDeclarativeConfigFromSpec;
